source/air/src/printer.rs

```diff
@@ -495,3 +495,166 @@ impl NodeWriter {
 pub(crate) fn node_to_string(node: &Node) -> String {
     NodeWriter::new().node_to_string_indent(&"".to_string(), node)
 }
+
+pub fn singular_query_to_node(vars: Vec<String>, enss: Vec<Expr>, reqs: Vec<Expr>) -> Node {
+    let pp = Printer::new(false);
+    let mut nodes = Vec::new();
+    let mut var_nodes = Vec::new();
+    nodes.push(str_to_node("singular-check-valid"));
+    let ens = pp.exprs_to_node(&Arc::new(enss));
+    let req = pp.exprs_to_node(&Arc::new(reqs));
+    for v in vars {
+        var_nodes.push(Node::Atom(v));
+    }
+    nodes.push(Node::List(var_nodes));
+    nodes.push(ens);
+    nodes.push(req);
+    Node::List(nodes)
+}
+
+use std::collections::HashMap;
+pub(crate) fn expr_to_singular(
+    expr: &Expr,
+    tmp_idx: &mut u32,
+    fmap: &mut HashMap<Node, String>,
+) -> String {
+    match &**expr {
+        ExprX::Const(Constant::Nat(n)) => n.to_string(),
+        ExprX::Var(x) => x.to_string(),
+        ExprX::Binary(BinaryOp::EuclideanMod, lhs, rhs) => {
+            // x % y ->  x - y*tmp
+            let t = format!("tmp_{}", tmp_idx.to_string());
+            *tmp_idx += 1;
+            let s1 = expr_to_singular(lhs, tmp_idx, fmap);
+            let s2 = expr_to_singular(rhs, tmp_idx, fmap);
+            format!("(({}) - (({})*({})))", s1, s2, t)
+        }
+        ExprX::Binary(BinaryOp::Eq, lhs, rhs) => {
+            let s1 = expr_to_singular(lhs, tmp_idx, fmap);
+            let s2 = expr_to_singular(rhs, tmp_idx, fmap);
+            format!("({}) - ({})", s1, s2)
+        }
+        ExprX::Multi(op, exprs) => {
+            let mut ss = vec![];
+            let sop = match op {
+                MultiOp::Add => ") + (",
+                MultiOp::Sub => ") - (",
+                MultiOp::Mul => ") * (",
+                _ => panic!("unsupported singular multi op"),
+            };
+            for e in &**exprs {
+                ss.push(expr_to_singular(&e, tmp_idx, fmap));
+            }
+            format!("(({}))", ss.join(sop))
+        }
+        ExprX::Apply(fname, exprs) => {
+            if "Mul" == (**fname).as_str() {
```

Comment:
maybe just do the translation in place, same applies to `EucMod`.

Reply:
That's a good point, but after changing codes in that way, I just learned that constant multiplication still ends in that old path. So I think I will keep it this way.